src/mount.ts

```diff
@@ -42,35 +47,127 @@ interface MountingOptions<Props> {
   shallow?: boolean
 }
 
-// TODO improve the typings of the overloads
-
-type ExtractComponent<T> = T extends { new (): infer PublicInstance }
-  ? PublicInstance
-  : any
+export type ComputedOptions = Record<
+  string,
+  ((ctx?: any) => any) | WritableComputedOptions<any>
+>
+export type ObjectEmitsOptions = Record<
+  string,
+  ((...args: any[]) => any) | null
+>
+export type EmitsOptions = ObjectEmitsOptions | string[]
 
 // Component declared with defineComponent
 export function mount<TestedComponent extends ComponentPublicInstance>(
   originalComponent: { new (): TestedComponent } & Component,
-  options?: MountingOptions<TestedComponent['$props']>
+  options?: MountingOptions<TestedComponent['$props'], TestedComponent['$data']>
 ): VueWrapper<TestedComponent>
-// Component declared with { props: { ... } }
-export function mount<TestedComponent extends ComponentOptionsWithObjectProps>(
-  originalComponent: TestedComponent,
-  options?: MountingOptions<ExtractPropTypes<TestedComponent['props'], false>>
-): VueWrapper<ExtractComponent<TestedComponent>>
-// Component declared with { props: [] }
-export function mount<TestedComponent extends ComponentOptionsWithArrayProps>(
-  originalComponent: TestedComponent,
-  options?: MountingOptions<Record<string, any>>
-): VueWrapper<ExtractComponent<TestedComponent>>
+
+// Functional Component declared with (props, ctx)=> render
+export function mount<Props, RawBindings = object>(
+  setup: (
+    props: Readonly<Props>,
+    ctx: SetupContext
+  ) => RawBindings | RenderFunction,
+
+  options: MountingOptions<Props, {}>
```

Comment:
nit: as `{}` is the default value for the `Data` type, I think you can simplify to `MountingOptions<Props>`?